rs/execution_environment/src/execution_environment/tests.rs

```diff
@@ -4393,3 +4393,155 @@ fn cannot_accept_cycles_after_replying() {
         initial_cycles + (transferred_cycles / 2u64)
     );
 }
+
+const REPLY_REJECT_CLEANUP_CALLBACK_WAT: &str = r#"
+    (module
+        (import "ic0" "call_new"
+            (func $ic0_call_new
+                (param i32 i32)
+                (param $method_name_src i32)    (param $method_name_len i32)
+                (param $reply_fun i32)          (param $reply_env i32)
+                (param $reject_fun i32)         (param $reject_env i32)
+            ))
+        (import "ic0" "call_perform" (func $ic0_call_perform (result i32)))
+
+        (import "ic0" "trap" (func $ic0_trap (param i32) (param i32)))
+        (import "ic0" "call_on_cleanup" (func $ic0_call_on_cleanup (param i32) (param i32)))
+        (import "ic0" "msg_reject_code" (func $ic0_msg_reject_code (result i32)))
+
+        (import "ic0" "msg_reply" (func $msg_reply))
+        (import "ic0" "msg_reply_data_append"
+            (func $msg_reply_data_append (param i32) (param i32)))
+        
+        (func $dummy
+            (call $msg_reply_data_append
+                (i32.const 300) (i32.const 1))  ;; refers to 9 on the heap
+            (call $msg_reply)
+        )   
+
+
+        (func $test
+            (call $ic0_call_new
+                (i32.const 100) (i32.const 10)   ;; callee canister id = 0
+                (i32.const 0) (i32.const 4)      ;; refers to "test" on the heap
+                (i32.const 0) (i32.const 200)    ;; on_reply closure at table index 0
+                (i32.const 1) (i32.const 200))   ;; on_reject closure at table index 1
+            (call $ic0_call_on_cleanup 
+                (i32.const 2) (i32.const 200))   ;; cleanup closure at table index 2
+            (drop (call $ic0_call_perform))
+        )
+
+        (func $on_reply (param i32)
+            (call $ic0_trap 
+                (i32.const 200) (i32.const 12))  ;; reply callback traps
+        )
+
+        (func $on_reject (param i32) 
+            (call $ic0_trap 
+                (i32.const 200) (i32.const 12))  ;; reject callback traps
+        )
+
+        (func $on_cleanup (param i32)
+            i32.const 300
+            call $ic0_msg_reject_code        ;; cleanup can't reply, so
+            i32.store8                       ;; we write cleanup code to memory and retrieve it via $dummy
+        )
+
+        (export "canister_update test" (func $test))
+        (export "canister_update dummy" (func $dummy))
+        (memory $memory 1)
+        (export "memory" (memory $memory))
+        (data (i32.const 0) "test")
+        (data (i32.const 100) "\00\00\00\00\00\00\00\00\01\01") ;; cansister_id of the installed canister is 0
+        (data (i32.const 200) "trap message")
+        (data (i32.const 300) "\09")
+        (table 3 3 funcref)
+        (elem (i32.const 0) $on_reply $on_reject $on_cleanup)
+    )
+"#;
+
+#[test]
+fn can_access_reject_code_in_cleanup_call_rejected() {
```

Comment:
System API tests typically live in `rs/execution_environment/tests/hypervisor.rs`: could you please move your tests there?